src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+    let expand = expand_range.map(|_| ArrayIndexOp::range_full());
+
+    let mut expanded = Vec::with_capacity(ndim);
+    expanded.extend_from_slice(prefix);
+    expanded.extend(expand);
+    expanded.extend_from_slice(suffix);
+
+    Cow::Owned(expanded)
+}
+
+// See `mlx_get_item` in python/src/indexing.cpp and `getItem` in
+// mlx-swift/Sources/MLX/MLXArray+Indexing.swift
+fn get_item(src: &Array, index: impl ArrayIndex, stream: StreamOrDevice) -> Array {
+    use ArrayIndexOp::*;
+
+    match index.index_op() {
+        Ellipsis => src.clone(),
+        TakeIndex { index } => get_item_int(src, index, 0, stream),
+        TakeArray { indices } => get_item_array(src, &indices, 0, stream),
+        Slice(range) => get_item_slice(src, range, stream),
+        ExpandDims => unsafe {
+            // SAFETY: 0 is always a valid axis
+            src.expand_dims_device_unchecked(&[0], stream)
+        },
+    }
+}
+
+// See `mlx_get_item_nd` in python/src/indexing.cpp and `getItemNd` in
+// mlx-swift/Sources/MLX/MLXArray+Indexing.swift
+fn get_item_nd(src: &Array, operations: &[ArrayIndexOp], stream: StreamOrDevice) -> Array {
+    use ArrayIndexOp::*;
+
+    let mut src = Cow::Borrowed(src);
+
+    // The plan is as follows:
+    // 1. Replace the ellipsis with a series of slice(None)
+    // 2. Loop over the indices and calculate the gather indices
+    // 3. Calculate the remaining slices and reshapes
+
+    let operations = expand_ellipsis_operations(src.ndim(), operations);
+
+    // Gather handling
+
+    // compute gatherFirst -- this will be true if there is:
+    // - a leading array or index operation followed by
+    // - a non index/array (e.g. a slice)
+    // - an int/array operation
+    //
+    // - and there is at least one array operation (hanled below with haveArray)
+    let mut gather_first = false;
+    let mut have_array_or_index = false; // This is `haveArrayOrIndex` in the swift binding
+    let mut have_non_array = false;
+    for item in operations.iter() {
+        if item.is_array_or_index() {
+            if have_array_or_index && have_non_array {
+                gather_first = true;
+                break;
+            }
+            have_array_or_index = true;
+        } else {
+            have_non_array = have_non_array || have_array_or_index;
+        }
+    }
+
+    let array_count = operations.iter().filter(|op| op.is_array()).count();
+    let have_array = array_count > 0;
+
+    let mut remaining_indices: Vec<ArrayIndexOp> = Vec::new();
+    if have_array {
+        // apply all the operations (except for .newAxis) up to and including the
+        // final .array operation (array operations are implemented via gather)
+        let last_array_or_index = operations
+            .iter()
+            .rposition(|op| op.is_array_or_index())
+            .unwrap(); // safe because we know there is at least one array operation
+                       // let remaining = indices.split_off(last_array_or_index + 1);
```

Comment:
we can probably get rid of this right?

Reply:
it's a commented line:
`// let remaining = indices.split_off(last_array_or_index + 1);`

Reply:
> it's a commented line: `// let remaining = indices.split_off(last_array_or_index + 1);`

That line was supposed to go away. Somehow I didn't see that. But the `last_array_or_index` is used later

Reply:
yeah, it was just about the commented line

Reply:
Addressed in ec661bf335722cc03b7c0da398d6e683b0810bc1